File: dsa/ds/rust/src/queue.rs
struct Node<T> {
    value: T,
    next: Option<Box<Node<T>>>,
}

struct Queue<T> {
    head: Option<Box<Node<T>>>,
    tail: Option<Box<Node<T>>>,
    len: usize,
}

impl<T> Queue<T> {
    fn new() -> Queue<T> {
        Queue {
            head: None,
            tail: None,
            len: 0,
        }
    }
    fn enqueue(mut self, val: &T) {
        self.head = Some(Box::new(Node::<T> {
            value: val,
            next: None,
        }));
        self.tail = Some(Box::new(Node::<T> {
            value: val,
            next: None,
        }));
        self.len += 1;
    }
    fn dequeue(mut self) {
        match self.head {
            Some(ref node) => {
                self.head = node.next;
                self.len -= 1;
            }
            _ => {}
        }
    }

    fn peek(&self) -> Option<&T> {
        match self.head {
            Some(ref node) => Some(&node.value),
            None => None,
        }
    }
}

fn insert() {
    let q = Queue::<i32>::new();
    q.enqueue(1);
}

File: dsa/ds/go/queue/queue.go
package queue

type Node[T any] struct {
	Value T
	Next  *Node[T]
}

type Queue[T any] struct {
	Head *Node[T]
	Tail *Node[T]
	Size int
}

func New[T any]() *Queue[T] {
	return &Queue[T]{
		Size: 0,
	}
}

func (q *Queue[T]) Enqueue(value T) {
	q.Size++
	if q.Head == nil {
		q.Head = &Node[T]{Value: value}
		q.Tail = q.Head
		return
	}
	q.Tail.Next = &Node[T]{Value: value}
	q.Tail = q.Tail.Next
}

func (q *Queue[T]) Dequeue() *T {
	if q.Head == nil {
		return nil
	}
	q.Size--
	v := q.Head.Value
	q.Head = q.Head.Next
	if q.Size == 0 {
		q.Tail = nil
	}
	return &v
	// TODO: free memory
}

func (q *Queue[T]) Peak() T {
	return q.Head.Value
}
